FirebaseAuthUI/FUIPasswordSignInViewController.m

```diff
@@ -108,35 +109,65 @@ - (void)signInWithDefaultValue:(NSString *)email andPassword:(NSString *)passwor
-      [FIREmailAuthProvider credentialWithEmail:email password:password];
-  [self.auth signInAndRetrieveDataWithCredential:credential
-                                      completion:^(FIRAuthDataResult *_Nullable authResult,
-                                                   NSError *_Nullable error) {
-    [self decrementActivity];
-
-    if (error) {
-      switch (error.code) {
-        case FIRAuthErrorCodeWrongPassword:
-          [self showAlertWithMessage:FUILocalizedString(kStr_WrongPasswordError)];
-          return;
-        case FIRAuthErrorCodeUserNotFound:
-          [self showAlertWithMessage:FUILocalizedString(kStr_UserNotFoundError)];
-          return;
-        case FIRAuthErrorCodeUserDisabled:
-          [self showAlertWithMessage:FUILocalizedString(kStr_AccountDisabledError)];
-          return;
-        case FIRAuthErrorCodeTooManyRequests:
-          [self showAlertWithMessage:FUILocalizedString(kStr_SignInTooManyTimesError)];
+    [FIREmailAuthProvider credentialWithEmail:email password:password];
+
+    void (^completeSignInBlock)(FIRAuthDataResult *, NSError *) = ^(FIRAuthDataResult *authResult,
+                                                                    NSError *error) {
+      [self decrementActivity];
+
+      if (error) {
+        switch (error.code) {
+          case FIRAuthErrorCodeWrongPassword:
+            [self showAlertWithMessage:FUILocalizedString(kStr_WrongPasswordError)];
+            return;
+          case FIRAuthErrorCodeUserNotFound:
+            [self showAlertWithMessage:FUILocalizedString(kStr_UserNotFoundError)];
+            return;
+          case FIRAuthErrorCodeUserDisabled:
+            [self showAlertWithMessage:FUILocalizedString(kStr_AccountDisabledError)];
+            return;
+          case FIRAuthErrorCodeTooManyRequests:
+            [self showAlertWithMessage:FUILocalizedString(kStr_SignInTooManyTimesError)];
+            return;
+        }
+      }
+      [self.navigationController dismissViewControllerAnimated:YES completion:^{
+        [self.authUI invokeResultCallbackWithAuthDataResult:authResult error:error];
+      }];
+    };
+
+  // Check for the presence of an anonymous user and whether automatic upgrade is enabled.
+  if (self.auth.currentUser.isAnonymous &&
+    [FUIAuth defaultAuthUI].shouldAutoUpgradeAnonymousUsers) {
+
+    [self.auth.currentUser
+        linkAndRetrieveDataWithCredential:credential
+                               completion:^(FIRAuthDataResult *_Nullable authResult,
+                                            NSError *_Nullable error) {
+      if (error) {
+        if (error.code == FIRAuthErrorCodeEmailAlreadyInUse) {
+          NSDictionary *userInfo = @{ FUIAuthCredentialKey : credential };
+          NSError *mergeError = [NSError errorWithDomain:FUIAuthErrorDomain
+                                                    code:FUIAuthErrorCodeMergeConflict
+                                                userInfo:userInfo];
```

Comment:
nit: this could maybe be moved into `FUIAuthErrors.m`.

Reply:
Moved to FUIAuthErrorUntils.